### ISSUE TITLE
Alpha for quantile regression doesn't seem to work

### ISSUE THREAD

**Issue by elaser** [Microsoft/LightGBM#1109]
Hi there!
  I've been working with @ClimbsRocks on using [auto_ml](https://github.com/ClimbsRocks/auto_ml) as a wrapper on top of lightgbm for a lot of our use cases at work.  We're super excited that lightgbm supports quantile regression.  I saw earlier in #1036 that it looks like alpha issue was resolved, but when I pass in different alphas, I'm not seeing different results.  

On my end, I wrote some functions to check if there were any difference between passing in alpha=.86 and alpha=.95 and I keep on getting results that were ~ 58th percentile as opposed to 86th or 95th on my validation and test sets.

@ClimbsRocks wrote something to test this:
```
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt


## ADDED
import lightgbm as lgb
## END ADDED


from sklearn.ensemble import GradientBoostingRegressor

np.random.seed(1)


def f(x):
    """The function to predict."""
    return x * np.sin(x)

#----------------------------------------------------------------------
#  First the noiseless case
X = np.atleast_2d(np.random.uniform(0, 10.0, size=10000)).T
X = X.astype(np.float32)

# Observations
y = f(X).ravel()

dy = 1.5 + 1.0 * np.random.random(y.shape)
noise = np.random.normal(0, dy)
y += noise
y = y.astype(np.float32)

# Mesh the input space for evaluations of the real function, the prediction and
# its MSE
xx = np.atleast_2d(np.linspace(0, 10, 100000)).T
xx = xx.astype(np.float32)

alpha = 0.95

# clf = GradientBoostingRegressor(loss='quantile', alpha=alpha,
#                                 n_estimators=250, max_depth=3,
#                                 learning_rate=.1, min_samples_leaf=9,
#                                 min_samples_split=9)

# clf.fit(X, y)


## ADDED
clf = lgb.LGBMRegressor(objective='quantile',
                        alpha=alpha,
                        num_leaves=31,
                        learning_rate=0.05,
                        n_estimators=100)
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='quantile',
        early_stopping_rounds=5)
## END ADDED


# Make the prediction on the meshed x-axis
y_upper = clf.predict(xx)

# clf.set_params(alpha=1.0 - alpha)
# clf.fit(X, y)


## ADDED
clf.set_params(alpha=1.0 - alpha)
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='quantile',
        early_stopping_rounds=5)
## END ADDED


# Make the prediction on the meshed x-axis
y_lower = clf.predict(xx)

# clf.set_params(loss='ls')
# clf.fit(X, y)


## ADDED
clf.set_params(objective='regression', eval_metric='l2')
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='l2',
        early_stopping_rounds=5)
## END ADDED


# Make the prediction on the meshed x-axis
y_pred = clf.predict(xx)

# Plot the function, the prediction and the 90% confidence interval based on
# the MSE
fig = plt.figure()
plt.plot(xx, f(xx), 'g:', label=u'$f(x) = x\,\sin(x)$')
plt.plot(X, y, 'b.', markersize=10, label=u'Observations')
plt.plot(xx, y_pred, 'r-', label=u'Prediction')
plt.plot(xx, y_upper, 'k-')
plt.plot(xx, y_lower, 'k-')
plt.fill(np.concatenate([xx, xx[::-1]]),
         np.concatenate([y_upper, y_lower[::-1]]),
         alpha=.5, fc='b', ec='None', label='90% prediction interval')
plt.xlabel('$x$')
plt.ylabel('$f(x)$')
plt.ylim(-10, 20)
plt.legend(loc='upper left')
plt.show()
```

and was seeing something like:

![image](https://user-images.githubusercontent.com/2738335/33748661-a1a8b558-db7e-11e7-932e-153150929f06.png)

Is alpha hardcoded in somewhere or is this user error on my end?

**Comment by guolinke** [comment]
@wxchan @StrikerRUS can you help to debug this? I think the alpha parameter is not passed right in python package

**Comment by wxchan** [comment]
@guolinke I add a log to print out config.alpha here https://github.com/Microsoft/LightGBM/blob/8ebef94cfe0627aef6def578fb0e6a2e082dbf1e/src/metric/regression_metric.hpp#L143, I think alpha is successfully set.

**Comment by guolinke** [comment]
can you try following:
1. reduce the number of points, e.g. 1000
2. more iterations with larger learning rate

**Comment by elaser** [comment]
not seeing any improvements:



My parameters are 
```
{'num_leaves': 8, 'learning_rate': 0.3, 'lambda_l2': 0.001, 'n_estimators': 1000, 'num_iteration': 10, 'histogram_pool_size': 16384}
```

and I'm seeing results like: 
```
('Number of quantile estimate >= actual_offset: ', 369870)
('Number of quantile estimate < actual_offset: ', 267464)
```

It doesn't seem to be much different from the result of running non-quantile regression

**Comment by guolinke** [comment]
did you use the latest master code?

**Comment by elaser** [comment]
early stopping?  like not enough estimators?  I installed latest from master and still seeing similar results.

**Comment by guolinke** [comment]
@elaser 
just try it with 0.99 and 0.9 . the result is normal:
```
import numpy as np
import matplotlib.pyplot as plt


## ADDED
import lightgbm as lgb
## END ADDED


from sklearn.ensemble import GradientBoostingRegressor

np.random.seed(1)


def f(x):
    """The function to predict."""
    return x * np.sin(x)

#----------------------------------------------------------------------
#  First the noiseless case
X = np.atleast_2d(np.random.uniform(0, 10.0, size=10000)).T
X = X.astype(np.float32)

# Observations
y = f(X).ravel()

dy = 1.5 + 1.0 * np.random.random(y.shape)
noise = np.random.normal(0, dy)
y += noise
y = y.astype(np.float32)

# Mesh the input space for evaluations of the real function, the prediction and
# its MSE
xx = np.atleast_2d(np.linspace(0, 10, 10000)).T
xx = xx.astype(np.float32)

alpha = 0.99

# clf = GradientBoostingRegressor(loss='quantile', alpha=alpha,
#                                 n_estimators=250, max_depth=3,
#                                 learning_rate=.1, min_samples_leaf=9,
#                                 min_samples_split=9)

# clf.fit(X, y)


## ADDED
clf = lgb.LGBMRegressor(objective='quantile',
                        alpha=alpha,
                        num_leaves=31,
                        learning_rate=0.1,
                        n_estimators=1000)
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='quantile')
## END ADDED


# Make the prediction on the meshed x-axis
y_upper = clf.predict(xx)

# clf.set_params(alpha=1.0 - alpha)
# clf.fit(X, y)


## ADDED
clf.set_params(alpha=1.0 - alpha)
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='quantile')
## END ADDED


# Make the prediction on the meshed x-axis
y_lower = clf.predict(xx)

# clf.set_params(loss='ls')
# clf.fit(X, y)


## ADDED

alpha=0.9
clf.set_params(alpha=alpha)
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='quantile')
## END ADDED


# Make the prediction on the meshed x-axis
y_upper_90 = clf.predict(xx)

# clf.set_params(alpha=1.0 - alpha)
# clf.fit(X, y)


## ADDED
clf.set_params(alpha=1.0 - alpha)
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='quantile')
## END ADDED


# Make the prediction on the meshed x-axis
y_lower_10 = clf.predict(xx)


## ADDED
clf.set_params(objective='regression', eval_metric='l2')
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='l2',
        early_stopping_rounds=5)
## END ADDED


# Make the prediction on the meshed x-axis
y_pred = clf.predict(xx)

# Plot the function, the prediction and the 90% confidence interval based on
# the MSE
fig = plt.figure()
plt.plot(xx, f(xx), 'g:', label=u'$f(x) = x\,\sin(x)$')
plt.plot(X, y, 'b.', markersize=10, label=u'Observations')
plt.plot(xx, y_pred, 'r-', label=u'Prediction')
plt.plot(xx, y_upper, 'k-')
plt.plot(xx, y_lower, 'k-')


plt.plot(xx, y_upper_90, 'y-')
plt.plot(xx, y_lower_10, 'y-')


plt.xlabel('$x$')
plt.ylabel('$f(x)$')
plt.ylim(-10, 20)
plt.legend(loc='upper left')
plt.show()
```

![image](https://user-images.githubusercontent.com/16040950/33821863-66733354-de90-11e7-958a-cfae82a20d53.png)

**Comment by guolinke** [comment]
for the quantile, you need to run with much more iteration. it is hard to convergence. 
you also can set ```reg_sqrt=True" to speed up the convergence.
```
import numpy as np
import matplotlib.pyplot as plt


## ADDED
import lightgbm as lgb
## END ADDED


from sklearn.ensemble import GradientBoostingRegressor

np.random.seed(1)


def f(x):
    """The function to predict."""
    return x * np.sin(x)

#----------------------------------------------------------------------
#  First the noiseless case
X = np.atleast_2d(np.random.uniform(0, 10.0, size=10000)).T
X = X.astype(np.float32)

# Observations
y = f(X).ravel()

dy = 1.5 + 1.0 * np.random.random(y.shape)
noise = np.random.normal(0, dy)
y += noise
y = y.astype(np.float32)

# Mesh the input space for evaluations of the real function, the prediction and
# its MSE
xx = np.atleast_2d(np.linspace(0, 10, 10000)).T
xx = xx.astype(np.float32)

alpha = 0.99

# clf = GradientBoostingRegressor(loss='quantile', alpha=alpha,
#                                 n_estimators=250, max_depth=3,
#                                 learning_rate=.1, min_samples_leaf=9,
#                                 min_samples_split=9)

# clf.fit(X, y)


## ADDED
clf = lgb.LGBMRegressor(objective='quantile',
                        alpha=alpha,
                        num_leaves=31,
                        learning_rate=0.1,
                        n_estimators=100,
                        reg_sqrt=True)
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='quantile')
## END ADDED


# Make the prediction on the meshed x-axis
y_upper = clf.predict(xx)

# clf.set_params(alpha=1.0 - alpha)
# clf.fit(X, y)


## ADDED
clf.set_params(alpha=1.0 - alpha)
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='quantile')
## END ADDED


# Make the prediction on the meshed x-axis
y_lower = clf.predict(xx)

# clf.set_params(loss='ls')
# clf.fit(X, y)


## ADDED

alpha=0.9
clf.set_params(alpha=alpha)
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='quantile')
## END ADDED


# Make the prediction on the meshed x-axis
y_upper_90 = clf.predict(xx)

# clf.set_params(alpha=1.0 - alpha)
# clf.fit(X, y)


## ADDED
clf.set_params(alpha=1.0 - alpha)
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='quantile')
## END ADDED


# Make the prediction on the meshed x-axis
y_lower_10 = clf.predict(xx)


## ADDED
clf.set_params(objective='regression', eval_metric='l2')
clf.fit(X, y,
        eval_set=[(X, y)],
        eval_metric='l2',
        early_stopping_rounds=5)
## END ADDED


# Make the prediction on the meshed x-axis
y_pred = clf.predict(xx)

# Plot the function, the prediction and the 90% confidence interval based on
# the MSE
fig = plt.figure()
plt.plot(xx, f(xx), 'g:', label=u'$f(x) = x\,\sin(x)$')
plt.plot(X, y, 'b.', markersize=10, label=u'Observations')
plt.plot(xx, y_pred, 'r-', label=u'Prediction')
plt.plot(xx, y_upper, 'k-')
plt.plot(xx, y_lower, 'k-')


plt.plot(xx, y_upper_90, 'y-')
plt.plot(xx, y_lower_10, 'y-')


plt.xlabel('$x$')
plt.ylabel('$f(x)$')
plt.ylim(-10, 20)
plt.legend(loc='upper left')
plt.show()
```

![image](https://user-images.githubusercontent.com/16040950/33822041-20b1da22-de91-11e7-9392-d5b1856b804e.png)

**Comment by elaser** [comment]
thanks @guolinke  -- i bumped up the learning rate and added `reg_sqrt=True` and that made all the difference!

**Comment by joseortiz3** [comment]
Why do the eval_metric and objective both need to be set to `quantile`? Shouldn't setting one automatically set the other?

**Comment by Laurae2** [comment]
@joseortiz3 because you might want to observe something else than quantile loss when performing quantile regression (such as mse, mae, etc. or a self-defined business metric). For a quantile task, in most cases you will watch the MSE metric and not the quantile metric.

**Comment by guolinke** [comment]
@joseortiz3 
The main difference is the leaf_output. Sklearn use the according percentile value as output, while LightGBM uses gradients directly. 
As a result, you can use larger learning_rate, e.g. 0.3 or 0.4. 
You also can set ```reg_sqrt``` to ```True``` to speed up the convergence.

**Comment by joseortiz3** [comment]
It's interesting that `reg_sqrt` helps, since it is just one type of one-to-one transformation one could apply to the supervised labels. This reminded me of the importance of making sure your prediction target is nicely distributed - in my case I think most of the slow convergence comes from the extremely skewed distribution of stock volumes. It did much better after I applied reg_sqrt, but even better once I applied log(1 + label / constant). Sqrt is just one example of a one-to-one mapping that squishes fat-tailed distributions closer to the origin. Thanks.

**Comment by guolinke** [comment]
@joseortiz3 
Yeah, the reason the calculation of leaf-output.

LightGBM's output is based on objective function (quantile loss), whose gradients are fixed values:
https://github.com/Microsoft/LightGBM/blob/master/src/objective/regression_objective.hpp#L395-L401 .
And the leaf-output is sum_grad / sum_hess. As a result, its outputs have no relation with the label magnitude.  So using `reg_sqrt` can help convergence, since it reduces range of labels.

For the sklearn, though its gradient is also based on quantile loss. However,
its output is based on label(y) directly. So it will be much faster. : https://github.com/scikit-learn/scikit-learn/blob/c9e6d4d3a42b966b5f00725a0fce52c5517c933e/sklearn/ensemble/gradient_boosting.py#L434-L442 

`quantile_l2` uses similar strategy, it changes the quantile loss to L2 form, as a result, the gradients will be related to the magnitude of labels. : https://github.com/Microsoft/LightGBM/blob/master/src/objective/regression_objective.hpp#L443-L449
